dcontext/debug_test.go

```diff
@@ -12,17 +12,39 @@ import (
 )
 
 func TestDebug(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	baseStr := fmt.Sprint(ctx)
-	ctx = dcontext.WithSoftness(ctx)
-
-	softStr := fmt.Sprint(ctx)
-	hardStr := fmt.Sprint(dcontext.HardContext(ctx))
-
-	assert.Truef(t, strings.HasPrefix(softStr, baseStr+"."),
-		"Soft Context string does not start with %q: %q", baseStr+".", softStr)
-	assert.Equal(t,
-		softStr+".HardContext",
-		hardStr)
+	ctx := context.Background()
+	var cancel context.CancelFunc
+	testcases := []struct {
+		Name   string
+		SetCtx func()
+		Suffix string
+	}{
+		{
+			Name: "WithCancel",
+			SetCtx: func() {
+				ctx, cancel = context.WithCancel(ctx)
+				t.Cleanup(cancel)
+			},
+			Suffix: "WithCancel",
+		},
+		{"WithSoftness", func() { ctx = dcontext.WithSoftness(ctx) }, ""},
+		{"HardContext", func() { ctx = dcontext.HardContext(ctx) }, "HardContext"},
+		{"WithoutCancel", func() { ctx = dcontext.WithoutCancel(ctx) }, "WithoutCancel"},
+	}
+	t.Log(fmt.Sprint(ctx))
+	for _, tc := range testcases {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
+			before := fmt.Sprint(ctx)
+			tc.SetCtx()
```

Comment:
There's only one ctx variable declared at top level in the `TestDebug()` function so this will use the `ctx` from the previous test. Is that intentional?

Reply:
Yes, it is actually required.  I suppose I should add a comment about what's going on.